### PR TITLE
 *: update stats using query feedback

### PR DESCRIPTION
**Pull request by lamxTyler**
This pr dumps the collected feedback into kv and use these feedback to update the stats, including the histogram and CM Sketch.
PTAL @coocood @winoros @shenli

### CODE REVIEW

File: domain/domain.go [hunk 1]
```diff
@@ -653,32 +657,45 @@ func (do *Domain) updateStatsWorker(ctx sessionctx.Context, owner owner.Manager)
...
+			err = statsHandle.LoadNeededHistograms()
 			if err != nil {
 				log.Error("[stats] load histograms fail: ", errors.ErrorStack(err))
 			}
+		case <-loadFeedbackTicker.C:
+			if !owner.IsOwner() {
+				continue
+			}
+			err = statsHandle.HandleUpdateStats()
+			if err != nil {
```
Comment (shenli) [inline domain/domain.go]: Add a metrics here.

File: domain/domain.go [hunk 2]
```diff
@@ -653,32 +657,45 @@ func (do *Domain) updateStatsWorker(ctx sessionctx.Context, owner owner.Manager)
...
+				log.Errorf("[stats] update stats using feedback fail: ", errors.ErrorStack(err))
+			}
+		case <-dumpFeedbackTicker.C:
+			err = statsHandle.DumpStatsFeedbackToKV()
+			if err != nil {
```
Comment (shenli) [inline domain/domain.go]: ditto

File: statistics/cmsketch.go [hunk 3]
```diff
@@ -54,6 +54,17 @@ func (c *CMSketch) InsertBytes(bytes []byte) {
 	}
 }
 
+func (c *CMSketch) setValue(h1, h2 uint64, count uint32) {
```
Comment (shenli) [inline statistics/cmsketch.go]: Add comment for this function.

File: statistics/feedback.go [hunk 4]
```diff
@@ -493,3 +493,123 @@ func buildNewHistogram(h *Histogram, buckets []bucket) *Histogram {
...
+	var buf bytes.Buffer
+	enc := gob.NewEncoder(&buf)
+	err = enc.Encode(pb)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return buf.Bytes(), nil
+}
+
+func decodeFeedback(val []byte, q *QueryFeedback, c *CMSketch) error {
```
Comment (shenli) [inline statistics/feedback.go]: Add comments for the following code logic.

File: statistics/update_test.go [hunk 5]
```diff
@@ -446,9 +446,24 @@ func (s *testStatsUpdateSuite) TestQueryFeedback(c *C) {
 		if t.idxCols == 0 {
 			c.Assert(feedback[0].DecodeInt(), IsNil)
 		}
-		c.Assert(statistics.UpdateHistogram(feedback[0].Hist(), feedback).ToString(t.idxCols), Equals, t.hist)
+		c.Assert(statistics.UpdateHistogram(feedback[0].Hist(), feedback[0]).ToString(t.idxCols), Equals, t.hist)
 	}
 
+	for _, t := range tests {
```
Comment (shenli) [inline statistics/update_test.go]: Add comment for the test case.

File: statistics/update.go [hunk 6]
```diff
@@ -190,6 +191,100 @@ func (h *Handle) dumpTableStatDeltaToKV(id int64, delta variable.TableDelta) (bo
 	return updated, errors.Trace(err)
 }
 
+// DumpStatsFeedbackToKV dumps the stats feedback to KV.
```
Comment (shenli) [inline statistics/update.go]: Do we need to add a few metrics for the following operations?

File: domain/domain.go (hunk 1 above)
Comment (lamxTyler) [reply domain/domain.go]: I think it is better to do add metrics in a separate pr, it is already too large now.

File: statistics/update.go (hunk 6 above)
Comment (lamxTyler) [reply statistics/update.go]: Yes, we need to.

File: domain/domain.go (hunk 2 above)
Comment (lamxTyler) [reply domain/domain.go]: I will do it in a separate pr.

File: statistics/update.go [hunk 7]
```diff
@@ -190,6 +191,115 @@ func (h *Handle) dumpTableStatDeltaToKV(id int64, delta variable.TableDelta) (bo
...
+	return errors.Trace(err)
+}
+
+func (h *Handle) dumpStatsUpdateToKV(tableID int64, isIndex int, q *QueryFeedback, hist *Histogram, cms *CMSketch) error {
+	hist = UpdateHistogram(hist, q)
+	err := SaveStatsToStorage(h.ctx, tableID, -1, isIndex, hist, cms)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	sql := fmt.Sprintf("delete from mysql.stats_feedback where table_id = %d and hist_id = %d and is_index = %d", tableID, hist.ID, isIndex)
```
Comment (coocood) [inline statistics/update.go]: May the delete exceeds transaction limit?

File: statistics/feedback.go [hunk 8]
```diff
@@ -493,3 +493,127 @@ func buildNewHistogram(h *Histogram, buckets []bucket) *Histogram {
 	}
 	return hist
 }
+
+// QueryFeedbackPB is used to serialize the QueryFeedback.
+type QueryFeedbackPB struct {
```
Comment (coocood) [inline statistics/feedback.go]: The name has `PB` but It's not `protobuf`.
Do you plan to change it to `protobuf` in the future?

File: executor/aggregate_test.go [hunk 9]
```diff
@@ -243,7 +243,7 @@ func (s *testSuite) TestAggregation(c *C) {
 
 	result = tk.MustQuery("select count(*) from information_schema.columns")
 	// When adding new memory columns in information_schema, please update this variable.
-	columnCountOfAllInformationSchemaTables := "737"
+	columnCountOfAllInformationSchemaTables := "741"
```
[the remark this answers is not in the document]
Comment (lamxTyler) [reply executor/aggregate_test.go]: I think it is 4 column added.

**Comment by coocood** [comment]
LGTM

File: statistics/cmsketch.go [hunk 10]
```diff
@@ -54,6 +54,18 @@ func (c *CMSketch) InsertBytes(bytes []byte) {
 	}
 }
 
+// setValue sets the count for value that hashed into (h1, h2).
```
[the remark this answers is not in the document]
Comment (lamxTyler) [reply statistics/cmsketch.go]: No, it means the hash value pair.

**Comment by lamxTyler** [comment]
PTAL @winoros

File: executor/aggregate_test.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (winoros) [reply executor/aggregate_test.go]: Yeah, i misread it.

File: statistics/feedback.go [hunk 11]
```diff
@@ -493,3 +493,127 @@ func buildNewHistogram(h *Histogram, buckets []bucket) *Histogram {
...
+		}
+	}
+	pb.Counts = append(pb.Counts, pointCounts...)
+	return pb
+}
+
+func encodeFeedback(q *QueryFeedback) ([]byte, error) {
+	var pb *queryFeedback
+	var err error
+	if q.hist.tp.Tp == mysql.TypeLong {
```
Comment (winoros) [inline statistics/feedback.go]: Can we make sure that the Tp is only `TypeLong`? Does it possible to be `TypeLongLong`?

File: statistics/update.go [hunk 12]
```diff
@@ -190,6 +191,116 @@ func (h *Handle) dumpTableStatDeltaToKV(id int64, delta variable.TableDelta) (bo
...
+	}
+	sql := fmt.Sprintf("insert into mysql.stats_feedback (table_id, hist_id, is_index, feedback) values "+
+		"(%d, %d, %d, X'%X')", fb.tableID, fb.hist.ID, isIndex, vals)
+	_, err = h.ctx.(sqlexec.SQLExecutor).Execute(context.TODO(), sql)
+	return errors.Trace(err)
+}
+
+// HandleUpdateStats update the stats using feedback.
+func (h *Handle) HandleUpdateStats(is infoschema.InfoSchema) error {
+	sql := fmt.Sprintf("select table_id, hist_id, is_index, feedback from mysql.stats_feedback order by table_id, hist_id, is_index")
```
Comment (zz-jason) [inline statistics/update.go]: no need to `fmt.Sprintf`

File: statistics/update.go [hunk 13]
```diff
@@ -190,6 +191,116 @@ func (h *Handle) dumpTableStatDeltaToKV(id int64, delta variable.TableDelta) (bo
...
+	}
+	var isIndex int64
+	if fb.hist.tp.Tp == mysql.TypeLong {
+		isIndex = 0
+	} else {
+		isIndex = 1
+	}
+	sql := fmt.Sprintf("insert into mysql.stats_feedback (table_id, hist_id, is_index, feedback) values "+
+		"(%d, %d, %d, X'%X')", fb.tableID, fb.hist.ID, isIndex, vals)
+	_, err = h.ctx.(sqlexec.SQLExecutor).Execute(context.TODO(), sql)
```
Comment (zz-jason) [inline statistics/update.go]: why not using `RestrictedSQLExecutor` ?

**Comment by lamxTyler** [comment]
PTAL @winoros

**Pull request merged**